### PR TITLE
intel-oneapi-compilers/mpi: added some paths to intel-oneapi-mpi to make compilers and libs visible

### PR DESCRIPTION
**Pull request by frankwillmore**
added some paths to intel-oneapi-mpi to make compilers and libs visible with module

### CODE REVIEW

**Review by adamjstewart — DISMISSED** [review]
@rscohn2 can you review?

File: var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]:
```suggestion
            env.prepend_path('LIBRARY_PATH', dir)
```
_library_path does the _join_prefix

File: var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]:
```suggestion
            env.prepend_path('LD_LIBRARY_PATH', dir)
```
_ld_library_path does the _join_prefix

File: var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]:
```suggestion

```

glob not used?

**Review by rscohn2 — CHANGES_REQUESTED** [review]

**Comment by frankwillmore** [comment]
I added some environment vars to the generated module so that the compiler wrappers will be able to find things. Also, I'm patching all versions of libmpi.so to add an rpath for libfabric so that built executables will always be able to find libfabric. This introduces a build dependency on patchelf. 

@rscohn2 when you have a chance, please look it over. It seems to be working as expected, and I can run wrapper-built executables with a purged module environment.

File: var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: Did you find this necessary? There were a couple reasons I wouldn't be sure:

* most MPI providers just set `MPICC` in `setup_run_environment`
* according to https://github.com/spack/spack/pull/20732/files#r553942281, `intel-oneapi-mpi` uses `MPICH_CC` rather than `I_MPI_CC`

Comment (frankwillmore) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: According to the help message that the wrapper prints, e.g. for mpicc:
```
 24     echo "   1. Command line option:  -cc=<compiler_name>"
 25     echo "   2. Environment variable: I_MPI_CC (current value '$I_MPI_CC')"
 26     echo "   3. Environment variable: MPICH_CC (current value '$MPICH_CC')"
```
either will be accepted, with MPICH_CC having the lowest precedent. I don't have a specific requirement for one over the other, just that one of these sets of variables needs to be set in the module. 

Having them specified here, rather than in the dependent_build_environment, ensures that these wrappers will work correctly when invoked from outside of spack, and this is a use case we care about. As much as we like spack for building, sometimes we just want the compiler.

Comment (scheibelp) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: Thanks for the information! This all looks good then, although I'm curious why you didn't have to set `MPICC` etc.

Comment (frankwillmore) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: Thanks for looking over it. MPICC isn't set when sourcing setvars.sh either. What specifically consumes MPICC? CMake findMPI or something similar?

Comment (scheibelp) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: I don't see it as a variable in https://cmake.org/cmake/help/latest/module/FindMPI.html, but I thought generally `MPICC` is how dependent packages could find the `mpicc` wrapper (and generally `MPICH_CC`/`I_MPI_CC` was how the `mpicc` wrapper found the underlying compiler). Perhaps CMake can locate the `mpicc` wrapper with `I_MPI_root`, but the Spack `openmpi` package for example sets `MPICC` in the environment.

Comment (rscohn2) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: I don't understand why we would set I_MPI_CC in this package. Shouldn't the compiler package set the variables? What if you want to use intel mpi wrappers and gcc as the underlying compiler?

Comment (frankwillmore) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: Good point. I see two solutions: 

- Set the variable conditionally when spec contains '%oneapi'
- Set the variable with the compiler

I'm leaning toward setting with the compiler. For us, this will eventually form part of a Cray PrgEnv, and having the oneapi compiler loaded would be a good reason to set this value.

Comment (rscohn2) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: Not knowing spack, I am still confused. In an earlier commit I added setup_dependent_build_environment for the MPI package to set MPICH_CC to spack_cc. I thought this would make the wrapper use the compiler that was part of the spec. Is this accurate? Is a case not covered?

Comment (scheibelp) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: > I added setup_dependent_build_environment for the MPI package to set MPICH_CC to spack_cc.

`setup_run_environment` runs under different conditions than `setup_dependent_build_environment` (EDIT): `setup_run_environment` is used to generate the module file and to set environment variables when you run `spack env activate ...`. So regarding 

> I thought this would make the wrapper use the compiler that was part of the spec. Is this accurate?

Yes

The changes in this PR are specifically to address https://github.com/spack/spack/pull/20808/files#r562032811

> [we want to ensure] that these wrappers will work correctly when invoked from outside of spack

Comment (frankwillmore) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: I moved the I_MPI_{CC,CXX,FC} environment settings to the intel-oneapi-compilers module as this seems to make the most sense for our expected usage patterns.

File: var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:86]: This is an interesting choice although it doesn't match typical Spack behavior for setting up the MPI compiler wrappers: usually we leave that up to the MPI instance since it generally feels out-of-place in the compiler module. Furthermore, to get this correct for all possible MPI implementations, we'd have to set all the variables here (e.g. `OMPI_CC`, `MPICH_CC`...).

Do you know if there is a fallback to using `CC` (in which case that part of https://github.com/spack/spack/pull/20717  could be useful)?

Comment (frankwillmore) [reply var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:86]: Afaik Intel mpi is the only mpi that recognizes this set of vars, and it doesn't pick up CC. I'd like to set here because the module will be more bulletproof. If you think it doesn't follow slack design patterns, we can just write the modules by hand. Adding the rpath for libfabric is needed, though.

Comment (scheibelp) [reply var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:86]: I don't think anything can go wrong with this except in strange cases (e.g. if using an `intel-oneapi-mpi` package not built with `icx` and loading the `intel-oneapi-compilers` module, which could reasonably be considered out-of scope) and I agree that it is helpful in some cases, so I'm inclined to approve it, but I want to think on it a bit more.

**Comment by tldahlgren** [comment]
@rscohn2 @scheibelp Is this PR ready for merging?  Todd would like to make a point release and my impression is this PR should be included.

File: var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: This appears to be hardcoded (users could place these files anywhere). It looks like this might be a dependency of `intel-oneapi-mpi` that we aren't currently accounting for. Or is this intended to help with externally-installed `intel-oneapi-mpi`?

**Comment by scheibelp** [comment]
I think there is one minor thing remaining: https://github.com/spack/spack/pull/20808#discussion_r569058285

The setting of `I_MPI_CC` etc. seems like a good idea to me after thinking on it more.

**Review by rscohn2 — DISMISSED** [review]

File: var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py
Comment (frankwillmore) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: This looks to have been a temporary fix before I got the rpath set. An additional value for the path is set to ".../mpi/latest/libfabric/lib/prov" but I need to dig a bit and see where. The hardcoded path should go away.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: > This looks to have been a temporary fix before I got the rpath set. An additional value for the path is set to ".../mpi/latest/libfabric/lib/prov" but I need to dig a bit and see where. The hardcoded path should go away.

Do you plan to address the hard-coded path in this PR?

Comment (frankwillmore) [reply var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py]: Yes. It is innocuous as-is, but there shouldn't be a hard coded path and I will look through to remember why I did this, and verify that it's no longer necessary. If you need to merge for a release, it shouldn't hurt anything, otherwise I'd need a day or two to sort it out.

**Comment by tldahlgren** [comment]
@frankwillmore Can you please resolve the conflict?

@scheibelp This PR has 2 approvals.  Do you still have changes you want before it gets merged?

**Comment by scheibelp** [comment]
> scheibelp This PR has 2 approvals. Do you still have changes you want before it gets merged?

Yes: https://github.com/spack/spack/pull/20808#discussion_r569058285 (it sounds like this will be resolved in a day or two). However, I'm not aware of the timeline for the release (e.g. if it can be pushed to next week). Since there is a conflict I think it will be relatively easy to resolve the request and the sync issue at the same time.

**Comment by scheibelp** [comment]
Hi @frankwillmore FYI there is also a merge conflict that will have to be resolved before this is merged.

**Comment by frankwillmore** [comment]
Ok, I fixed the merge conflict and re-tested, looks OK.

**Review by scheibelp — APPROVED** [review]

**Pull request merged**
